# FIT5196 Assessment 1 Task 1
#### Student Name: Pichaphop Sunthornjittanon
#### Student ID: 31258301

Date: 31/08/21
 ##############################


Environment: Python 3.7.11 and Google Collab

Libraries used:
* re (Regular expression)
* os (Operating System)


## 1. Introduction
This assessment touches the very first step of analyzing textual data, i.e., extracting data
from semi-structured text files. Each
text file contains information about the tweets, i.e., “user name”, “user code”, “user
description”, “number of followers”, “whether or not the user account is verified”, “date
of the tweet”, and the “tweet text”. Your task is to extract the data from the text file and
transform the data into a XML format with the following elements:
1. users: this tag wraps all the users
2. user: this tag wraps all the tweets from a particular user and keeps the meta data for
each user such as number of followers, verified or not, user description etc. If a user has
multiple tweets, the meta data of the latest tweet (i.e., the tweet with the most
recent date) must be used.
3. Tweets: wraps all the tweets of a specific user
4. tweet: for each user, this tag represents the text of the user tweet


## 2.  Importing libraries 

In [1]:
# Import required libraries

# Use for Regular Expression
import re

# Use for Operating System
import os

## 3. Examining and loading data

In this section, we loaded and roughly explored the file from google drive with the file name 31258301_task1_input.txt, which will be further processed

In [2]:
# Mount Google Drive in the colab environment
from google.colab import drive
drive.mount('/content/drive') 

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
# Read the whole file
with open('/content/drive/Shareddrives/FIT5196-s2-2021-tutorials/Assessment 1/Task1 datasets/31258301_task1_input.txt','r') as infile:
    text = infile.read()

In [4]:
# print first 20 lines of the file
with open('/content/drive/Shareddrives/FIT5196-s2-2021-tutorials/Assessment 1/Task1 datasets/31258301_task1_input.txt','r') as infile:
    print('\n'.join([infile.readline().strip() for i in range(0, 20)])) 

$username.: LIFE TOKEN HOLDERS
$user_code.: 100000001
$udesc.: Every 40 seconds someone commits suicide.
‍
Life Token is a charity-orientated token on the Binance Smart Chain. Our goal is to aid Suicide Prevention charities
$followerNo.: 6769.0 $verified?.: False $tweet_date.: 2021-02-10 23:58:48
$tweet.: 😎 Today, that's this #Thursday, we will do a "🎬 Take 2" with our friend @LeoWandersleb, #Btc #wallet #security expe… https://t.co/go6aDgRml5
$username.: Mehidi Hasan
$user_code.: 100030713
$tweet_date.: 2021-07-05 16:46:47
$tweet_text.: @IoLunaland project is presented by a competent team that has good potential for the implementation of the objectives. The key to the success of this project
#presale #blockchain #cryptotrading #investing #cryptocurrency #bitcoin #trading #altcoin #cryptonews
$verified_user?.: False $No. followers.: 2022.0 $user_desc.: nan
$user_name.: conor prunty
$user_code.: 100052719
$user_desc.: Shit is about to go down when you see ‘ach, go tobann...’
$followerNo

In [5]:
# Read the sample input file
# with open('/content/sample_input.txt','r') as infile:
#     text = infile.read()

## 4. Extracting data using regular expression

In this section, we want to use regular expression to extract following data including user name, user code, user description, number of followers, whether or not the user account is verified, date of tweet and tweet text from the given text file.

Regular Expression Explaination

- We use re.findall function to write regular expression and extract text,which match the specified pattern returning into list of texts
- r" some regular expression" is used as a raw string to suppress actual meaning of escape characters, which help the code to be more readable
- The pattern of extracted data is $(field name) : (extracted data)
- Each attribute may have different field name such as user_name and uname
- Some fields consist of multiple line data, so we use flags = DOTALL in order to make '.' special character capture everything including new line
- (?:) is used for matching the pattern, but not captureing of the term in the blacket, while () is used for matching and capturing 


In this following section, each regular expression logic is explained below

1. user_name - the username data was followed by the field name that start with $u and consist of the word 'name' following by ' .: '. Therefore, all texts were captured after the field name until the end of the lines.  

2. user_code - the user code data was followed by the field name that start with $u and consist of the word 'code' following by ' .: ', which has 9 digits. Therefore, 9 digit numbers after the field name were captured.

3. user_description - the user description data was followed by the field name that start with $u and consist of the word 'desc' following by ' .: '. User description can be in multiple line so we use flags = re.DOTALL that make '.' special characters capture everything including new line until the condition is met, which is the new the field username or number of followers. 

4. number_of_followers - the number of followers data was followed by the field name that consists of the word 'follower' and end by ' .: '. Therefore, all texts were captured after the field name until the end of the lines or finding the character \$.  

5. verified - the verification data was followed by the field name that consists of the word 'verif' and end by ' .: '. Therefore, all texts were captured after the field name until the end of the lines or finding the character \$.  

6. date of the tweet - the date of the tweet was followed by the field name that starts with the word '$tweet_date' and ends by ' .: '. Therefore, all texts were captured after the field name until the end of the lines or finding the character \$.  

7. tweet text - the field tweet text always follow by tweet date, but before username or verfied fields. So we capture the text between these two fields using dotall flag. Moreover, the tweet text is in the last field in the text so we include the patern of the lastest tweet and combind to not latest tweets

In [6]:
### 1.Regular expression to extract user_name ###
user_name = re.findall(r"(?:\$u.*name\.: )(.*)",text)

### 2.Regular expression to extract user_code ###
user_code = re.findall(r'(?:\$u.*code\.: )(\d{9})',text)

### 3.Regular expression to extract user_decription ###
user_description = re.findall(r'\$u.*?desc.*?\.: (.*?)(?:\n)(?:\$u.*?name\.:|\$follower.*?\.: |\$No.*?\.)',text,flags= re.DOTALL)

### 4.Regular expression to extract number of followers ###
number_of_followers = re.findall(r'follower.*?\.: (.*?) (?:\$|\n)',text)

### 5.Regular expression to extract verified ###
verified = re.findall('verif.*?\.: (.*?) (?:\$|\n)',text)

### 6.Regular expression to extract date_of_the_tweet ###
date_of_the_tweet = re.findall('\$tweet_date.*?\.: (.*?)(?:\$|\n)', text)

### 7.Regular expression to extract tweet text
tweet_not_last = re.findall('\$tweet_date.*?\.: (?:.*?)(?:\$|\n)\$.*?.: (.*?)(?:\n)(?:\$u.*?name\.: |\$verif.*?\.:)', text,flags= re.DOTALL) 
tweet_last = re.findall('(?:\$tweet_date.*?\.: (?:.*?)(?:\$|\n)\$.*.: (.*))', text,flags= re.DOTALL) 
tweet_last = list(map(lambda x: x.strip('\n'),tweet_last))
tweet = tweet_not_last +  tweet_last


In [7]:
## Check length of each field list

# user_name
print('The length of user name list is ',len(user_name))

# user_code
print('The length of user code list is ',len(user_code))

# user_description
print('The length of user description list is ',len(user_description))

# number_of_followers
print('The length of number of followers list is ',len(number_of_followers))

# verified
print('The length of verified list is ',len(verified))

# date_of_the_tweet
print('The length of date of the tweet list is ',len(date_of_the_tweet))

# tweet
print('The length of tweet list is ',len(tweet))


The length of user name list is  12022
The length of user code list is  12022
The length of user description list is  12022
The length of number of followers list is  12022
The length of verified list is  12022
The length of date of the tweet list is  12022
The length of tweet list is  12022


## 5. Handling XML Special Characters 

Since we want to generate XML file as an output by extracting data from the given text file,  we need to deal with texts that contain XML special characters ,namely &amp;, &quot; ,&apos;,&gt;,&lt; by replacing them with \&amp; , \&quot; , \&apos; ,\&gt; ,\&lt;. By doing that, the handle_XML_char is defined which replace XML special characters. After that, Map is used to map handle_XML_char to all field lists

In [8]:
# Define function that replace XML special characters 
def handle_XML_char(x) :
  return x.replace('&','&amp;').replace('"','&quot;').replace("'",'&apos;').replace('>','&gt;').replace('<','&lt;')

In [9]:
# Map the handle_XML_char function to all fields
user_name = list(map(handle_XML_char,user_name))
user_code = list(map(handle_XML_char,user_code))
user_description = list(map(handle_XML_char,user_description))
number_of_followers = list(map(handle_XML_char,number_of_followers))
verified = list(map(handle_XML_char,verified))
date_of_the_tweet = list(map(handle_XML_char,date_of_the_tweet))
tweet = list(map(handle_XML_char,tweet))

##6. Creating  placeholders for metadata, tweet data and user number

In this section, we aim to create 3 placeholders by utilising list and dictionary data structure 

**1.metadata** 
(user_name, number_of_followers,verified,latest_tweet_date)

We use nested dictionary to place metadata inside since when the data is iterated, it can check at the existing key and latest tweet date to design whether to update meta data or not.

This data is updated at most recent tweet date and is used user_no as a key in the main dictionary with the attributes as keys in sub-dictionaries

*The example of the meta data format is shown belown*


{

**'100000001'**: 

{'latest_tweet_date': '2021-07-05 16:52:10',

  'number_of_followers': '7449.0',

  'user_description': 'Every 40 seconds someone commits suicide.\n\u200d\nLife Token is a\xa0charity-orientated\xa0token on the\xa0Binance Smart Chain. Our goal is to\xa0aid Suicide Prevention charities',

  'user_name': 'LIFE TOKEN HOLDERS',

  'verified': 'False'},

 **'100030713':** {'latest_tweet_date': '2021-07-05 16:46:47',

  'number_of_followers': '2022.0',

  'user_description': 'nan',

  'user_name': 'Mehidi Hasan',
  
  'verified': 'False'}

}

**2. Tweet data**

We use dictionary that contains list of tweet data since in each iteration, we can  add more user_id and tweet text if the existing dictionary does not contain that data. However,if the the user_id is already existed, append the tweet data in the list.

*The example of the tweet data format is shown belown*

{

**'100001281'**: ['#CenturyMarketNews on #Cryptocurrency: 7/7\nThings could get even more exciting should the upcoming #SEC verdict res… https://t.co/AUt9NOd6Bu'],


 **'100013247'**: [
   
   '[NEW] Coinbase Announces First Quarter 2021 Estimated Results : Verified Users of 56M, Trading Volume of $335B, Tot… https://t.co/qxT4zfRVrR',
  
  '[NEW] What&apos;s the average holdings of a r/CryptoCurrency user? Share your portfolio value anonymously... +… https://t.co/XHW6lCXw04',
 
  '[NEW] Which are the most delusional crypto supporters, in your opinion + https://t.co/LNthwnpDhe #bitcoin #btc #cryptocurrency #crypto #blockchain']

}

**3. User number**
To maintain the order of users, we create the list of user_id that append everytime in each iteration if the user is not in the list.

*The example of the user number format is shown belown*

['100000001',
 '100030713',
 '100052719',
 '100033219']






In [10]:
# Initialise empty dictionary
user_dict = dict()
tweet_dict = dict()
user_no_list = list()

### For any data point in the tweet data ###
for no_user in range(0,len(user_name)):

  ## 1.If user is in the list, ##
  if user_code[no_user] in user_dict.keys():

    # Extract the specific user data in dictionary
    old_user_data = user_dict[user_code[no_user]]

    # Pull only date of the tweet of existing data
    old_user_latest_tweet_date = old_user_data['latest_tweet_date']

    # Look at date of the tweet of new data
    new_user_tweet_date = date_of_the_tweet[no_user]
    
    # check if date of tweet is not a recent one
    if old_user_latest_tweet_date > date_of_the_tweet[no_user] :

      # Chcek 
      # print('usercode',user_code[no_user])
      # print('old data',old_user_latest_tweet_date)
      # print('new data',new_user_tweet_date)
      
      # Append tweet data into tweet dict in specific usercode
      tweet_dict[user_code[no_user]].append(tweet [no_user])


      # Not update meta data
      continue

  
  ## 2.If the user is not in the list, ##
  else :

    # add the user in the user_no_list
    user_no_list.append(user_code[no_user])

    # Create new key (usercode) in tweet dictionary with value of empty list
    tweet_dict[user_code[no_user]] = []

  ## After that ##
  # Add/update meta data
  user_dict[user_code[no_user]] = {'user_name' : user_name[no_user],
                                  'user_description' :user_description[no_user],
                                  'number_of_followers' :number_of_followers[no_user],
                                  'verified' :verified[no_user],
                                  'latest_tweet_date' :date_of_the_tweet[no_user]
  }

  # Append tweet data into tweet dict in specific usercode
  tweet_dict[user_code[no_user]].append(tweet [no_user])


    
    


In [11]:
# Check the length of 3 placeholders
print('Lenght of user dict is ', len(user_dict)) 
print('Lenght of tweet dict is ', len(tweet_dict)) 
print('Lenght of user no list is ', len(user_no_list)) 


Lenght of user dict is  7318
Lenght of tweet dict is  7318
Lenght of user no list is  7318


##7. Writing XML file

The following code generates the final output, which is in XML format by opening write mode and name the file to 31258301.xml. The rest of the code uses for loop to write the given xml with data that can be extracted from metadata and tweet data dictionaries from the previous section

In [12]:
# Write XML file

with open('31258301.xml','w') as file :

  file.write('<users>')

  for user_no in user_no_list:
    file.write('<user name="'+user_dict[user_no]['user_name']+'">')
    file.write('<verified_user>'+user_dict[user_no]['verified']+'</verified_user>')
    file.write('<user_description>'+user_dict[user_no]['user_description']+'</user_description>')
    file.write('<no_followers>'+user_dict[user_no]['number_of_followers']+'</no_followers>')
    file.write('<tweets>')
    for index in range(len(tweet_dict[user_no])):
      file.write('<tweet>'+tweet_dict[user_no][index]+'</tweet>') 
    file.write('</tweets>')
    file.write('</user>')
  file.write('</users>')


##Summary

In this assignment, we performed several steps to approach the problems and obtained the final result as the xml file name 31258301.xml. The steps included

1. Introduction - Understand the requirement of this assignment
2. Importing libraries -Import re and os libraries (only two are allowed)
3. Examining and loading data - Load data from the given text file
4. Extracting data using regular expression -Use regular expression to find the pattern and extract data from the text
5. Handling XML Special Characters - Replace special characeters
6. Creating placeholders for metadata, tweet data and user number - Create format of the data that will use for loop in the writing XML file part
7. Writing XML file - Generate the final output 31258301.xml